docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```diff
@@ -0,0 +1,304 @@
+
+Evidence has the following simple interface:
+
+```go
+type Evidence interface {  //proposed
+  Height() int64                                     // height of the equivocation
+  Bytes() []byte                                     // bytes which comprise the evidence
+  Hash() []byte                                      // hash of the evidence
+  ValidateBasic() error
+  String() string
+}
+```
+
+We have two recognized evidence that complies with it
+
+```go
+type LightClientAttackEvidence struct {
+  ConflictingBlock *LightBlock
+  CommonHeight int64
+}
+```
+where the `Hash()` is the hash of the header and commonHeight
+
+```go
+type DuplicateVoteEvidence {
+  VoteA *Vote
+  VoteB *Vote
+}
+```
+where the `Hash()` is the hash of the two votes
+
+The first is generated in the light client, the second in consensus. Both are sent to the evidence pool through `AddEvidence(ev Evidence) error` where it first runs `Has(ev Evidence)` to check if it has already received it (by comparing hashes) and then  `Verify(ev Evidence) error` before adding it to the database
+
+
+`Verify()` does the following:
+
+- Use the hash to see if we already have this evidence in our committed database.
+
+- Use the height to check if the evidence hasn't expired.
+
+- If it hasn't expired use the height to find the block header and confirm that the time hasn't also expired.
+
+- Then proceed with switch statement for each of the two evidence:
+
+For `DuplicateVote`:
+
+- Check that height, round, type and validator address are the same
+
+- Check that the Block ID is different
+
+- Check the look up table for addresses to make sure there already isn't evidence against this validator
+
+- Fetch the validator set and confirm that the address is in the set
+
+- Check that the chain ID and signature is valid.
+
+For `LightClientAttack`
+
+- Fetch the common signed header and val set from the common height and use skipping verification to verify the conflicting header
+
+- Fetch the trusted signed header at the same height as the conflicting header and compare with the conflicting header to work out which type of attack it is and in doing so return the malicious validators.
+
+  - If equivocation the validators that signed for the commits of both the trusted and signed header
+
+  - If lunatic the validators from the common val set that signed in the conflicting block
+
+  - If amnesia none (we can't know which validators are malicious). This also means that we don't
+  send amnesia evidence to the application
+
+- For each validator, check the look up table to make sure there already isn't evidence against this validator
+
+
+After verification we save the evidence with the following key `height/hash` and in the following struct
+
+```go
+type EvidenceInfo struct {
+  ev Evidence
+  time time.Time
+  validators []Validator
+  totalVotingPower int64
+}
+```
+
+We save extra information as this will be useful in forming the ABCI Evidence that we will come to later
```

Comment:
Does this mean that evidence effectively metamorphoses from `Evidence` (either `LightClientAttackEvidence` or `DuplicateVoteEvidence`) to `EvidenceInfo` to `ABCIEvidence` as it passes through this evidence flow?

Reply:
Yes Evidence is what is created during detection, gossiped and committed on chain. EvidenceInfo contains extra information and is stored in the evidence pool. ABCI Evidence is the evidence sent to the application

Reply:
![complete-metamorphosis](https://user-images.githubusercontent.com/773524/92262809-f4c4ad00-eedb-11ea-984c-978f0aefc792.gif)

Reply:
(I feel we may want somewhat more descriptive names for these phases.)

Reply:
I like this but I don't think it is quite so linear nor process-like

Reply:
Haha, ok. Is that because evidence can go back and forth between these various "phases"? Or is it because evidence objects can split and merge at various phases?

Reply:
It's like we use different data structures to represent the information that the parts need. The application only needs abci.Evidence (they don't need all the verification data), the light client only knows as much information as `LightClientAttackEvidence` which is still small to gossip and commit but large enough for the full nodes (evidence pool) to fill in the rest of the gaps by adding the information in `EvidenceInfo`

Reply:
Sure. Are those transformations non-linear though?

(I just want my evidence to be a beautiful butterfly!)

Reply:
![evidence_lifecycle](https://user-images.githubusercontent.com/27455974/92373062-1309f300-f0fe-11ea-8900-3aa75746f068.png)

Not as pretty as yours @tessr but how about this as a schematic